ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldStateDownloader.java

```diff
@@ -373,4 +403,58 @@ private synchronized void markDone() {
   private boolean isRootState(final BlockHeader blockHeader, final NodeDataRequest request) {
     return request.getHash().equals(blockHeader.getStateRoot());
   }
+
+  private class PersistNodeDataTask extends AbstractEthTask<Void> {
+
+    private final List<Task<NodeDataRequest>> batch;
+    private final BlockHeader header;
+
+    public PersistNodeDataTask(final BlockHeader header) {
+      super(metricsSystem);
+      this.header = header;
+      batch = new ArrayList<>();
+    }
+
+    @Override
+    protected void executeTask() {
+      while (!isDone()) {
+        try {
+          final Task<NodeDataRequest> task = requestsToPersist.poll(1, TimeUnit.SECONDS);
+          if (task != null) {
+            batch.clear();
+            batch.add(task);
+            requestsToPersist.drainTo(batch, 1000);
+            final Updater storageUpdater = worldStateStorage.updater();
+            batch.forEach(
+                taskToPersist -> {
+                  final NodeDataRequest request = taskToPersist.getData();
+                  request.persist(storageUpdater);
+                  queueChildRequests(request);
+                  taskToPersist.markCompleted();
+                  completedRequestsCounter.inc();
+                });
+            storageUpdater.commit();
+
+            if (pendingRequests.allTasksCompleted()) {
+              final Updater updater = worldStateStorage.updater();
+              updater.putAccountStateTrieNode(header.getStateRoot(), rootNode);
+              updater.commit();
+              markDone();
+              result.get().complete(null);
+            } else {
+              ethContext.getScheduler().scheduleSyncWorkerTask(() -> requestNodeData(header));
+            }
+          }
+        } catch (final InterruptedException ignore) {
+          Thread.currentThread().interrupt();
+        } catch (final RuntimeException e) {
+          LOG.error("Unexpected error while persisting world state", e);
```

Comment:
Consider logging some context for this exception so that it is easier to track down (eg header, etc)

Reply:
We don't really have any context here - there's only one world state download so the header is useless and we're attempting to persist about 394 data points - far too much to include in a log.